woocommerce/woocommerce-ios
# [Mobile Payments] Refresh order on payment start

## Description


We've had a few reports of duplicated payments for a single order using the iOS app with In-Person Payments. See pdfdoF-2HE-p2 for full details.

The issue has occurred when there's poor network conditions, usually manifesting with an unknown server error (Yosemite.ServerSidePaymentCaptureError error 0.) This is a retryable error, so the merchant tapped `Try again`, which resulted in a new payment intent and attempt. Both these intents succeed (despite the original error) so the order is paid for twice.

I've not been able to fully reproduce the issue, but by simulating errors we can get close.

On Android, the order is refreshed at the start of each payment attempt, which makes this less likely. If the order is refreshed and found to be paid, the second payment attempt would not be allowed. If the order refresh fails, the second payment attempt would not be allowed either. That way, on Android, we know that when we start the payment the order has not already been paid for.

This PR adds this refresh behaviour to the iOS app.

## Testing instructions


Set up Proxyman or Charles and set up (but deactivate) a breakpoint on the `GET orders` endpoint: `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/orders/*`

1. Create an order, or use simple payments, and go through to the payment flow.
2. Select `Card Reader` or `Tap to Pay on iPhone`
3. Tap a real card, or use another method to cause a payment failure (e.g. [error testing order amounts](https://stripe.com/docs/terminal/references/testing?locale=en-GB#physical-test-cards))
4. Wait for the retryable error to be shown
5. Activate the `GET orders` breakpoint
6. Tap `Try Collecting Again`
7. Edit the response of the first `order` request to have a different modified time (just a few seconds later is fine – this triggers a full update).
8. Observe that a `Checking order` message is shown in the app while this is ongoing.
9. Edit the response of the second `order` request to have a date in the `datePaid` field – it will be null by default, this signifies that the order is paid.
10. Observe that a non-retriable error is shown explaining that the order has already been paid
11. Tap `Dismiss`
12. Observe that you return to the Order screen, and it shows the paid order.


## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/2472348/6e995f36-2541-4634-8d9c-af91f6cefaf9



---
